mm2src/lp_ordermatch/simple_market_maker.rs

```diff
@@ -440,30 +521,60 @@ async fn prepare_order(
         .await?
         .ok_or_else(|| MmError::new(OrderProcessingError::AssetNotEnabled))?;
 
-    info!("balance for {} is {}", cfg.base, base_balance);
+    debug!("balance for {} is {}", cfg.base, base_balance);
 
     let mut calculated_price = rates.price * cfg.spread.clone();
-    info!("calculated price is: {}", calculated_price);
+    debug!("calculated price is: {}", calculated_price);
     if cfg.check_last_bidirectional_trade_thresh_hold.unwrap_or(false) {
-        calculated_price = vwap_calculator(calculated_price.clone(), ctx, &cfg).await?;
+        calculated_price = vwap_calculator(calculated_price.clone(), ctx, cfg).await?;
     }
 
-    let volume = match &cfg.balance_percent {
-        Some(balance_percent) => balance_percent * &base_balance,
-        None => MmNumber::default(),
+    let mut is_max = cfg.max.unwrap_or(false);
+
+    let volume = if let Some(max_volume) = &cfg.max_volume {
+        match max_volume {
+            VolumeSettings::Percentage(balance_percent) => {
+                if *balance_percent >= MmNumber::from(1) {
+                    is_max = true;
```

Comment:
```rust
enum PreparedOrderVolume {
    Exact(MmNumber),
    Max,
}
```

will allow avoiding invalid occasional returns like `volume > 0` and `max: true` concurrently.

Reply:
those check are more for dynamic behaviour and to force the max parameter of `update_maker_order` and `setprice` in order to avoid calculation problems

Let's take the following configuration:

```json
"ETC/FIRO": {
            "base": "ETC",
            "rel": "FIRO",
            "max_volume": {"percentage": "1"},
            "min_volume": {"usd": "5"},
            "spread": "1.02",
            "base_confs": 1,
            "base_nota": false,
            "rel_confs": 1,
            "rel_nota": false,
            "enable": true,
            "price_elapsed_validity": 45,
            "check_last_bidirectional_trade_thresh_hold": false
        }
```

Here the user wants to sell 100% of his balance, but max parameter is not specified

I would like that is implicit:

```json
"ETC/FIRO": {
            "base": "ETC",
            "rel": "FIRO",
            "max": true,
            "min_volume": {"usd": "5"},
            "spread": "1.02",
            "base_confs": 1,
            "base_nota": false,
            "rel_confs": 1,
            "rel_nota": false,
            "enable": true,
            "price_elapsed_validity": 45,
            "check_last_bidirectional_trade_thresh_hold": false
        }
```

and:

```json
"ETC/FIRO": {
            "base": "ETC",
            "rel": "FIRO",
            "max_volume": {"usd": "10"},
            "min_volume": {"usd": "5"},
            "spread": "1.02",
            "base_confs": 1,
            "base_nota": false,
            "rel_confs": 1,
            "rel_nota": false,
            "enable": true,
            "price_elapsed_validity": 45,
            "check_last_bidirectional_trade_thresh_hold": false
        }
```

If I have like 8 dollars of ETC that would also implicitly max at runtime

Not sure if it's possible to know that before querying the balance.

Regarding the balance_percent it's will be confusing to have `max` field at two places at the same time I believe.

Reply:
I see, we can keep the current code then, thanks!